polyseam/cndi
# [Enhancement]: Add Node Pool Labeling and Tainting Support for AKS, GKE, and EKS

Issue #

# Description

This pull request introduces the capability to apply labels and taints directly to node pools in AKS, GKE, and EKS through our Kubernetes cluster management API. This enhancement simplifies the configuration of node-specific behaviors and ensures that our node specifications align with best practices for workload separation and node allocation.

# Changes
AKS Support: Updated the AKS module to handle node labeling and tainting within the node pool definition.
GKE Support: Extended the GKE module to include labels and taints in the node pool configuration.
EKS Support: Modified the EKS module to support tainting and labeling as part of the node pool setup.

# Test Instructions




Create a node pool with a taint key called "testing" and a value of "mytaint"
```
    nodes:
      - name: x-node-group
        instance_type: t3.medium
        disk_size: 100
        count: 3
        labels: 
          testing: 'mytaint'
        taints: 
          - key: testing
            value: 'mytaint'
            effect: NoSchedule

```
then create a pod that is tolerant of that taint
```
apiVersion: v1
kind: Pod
metadata:
  name: my-tolerant-pod
spec:
  containers:
  - name: my-container
    image: nginx
  tolerations:
  - key: "testing"
    operator: "Equal"
    value: "mytaint"
    effect: "NoSchedule"
   ```
This pod should be scheduled on that nodepool 
Then create a pod that is intolerant of that taint
```
  my-intolerant-pod:
    apiVersion: v1
    kind: Pod
    metadata:
      name: my-intolerant-pod
    spec:
      containers:
      - name: my-container
        image: nginx
```
This pod shouldnt be scheduled on that nodepool 
# Code of Conduct

By submitting this Pull Request, you agree to follow our
[Code of Conduct](https://github.com/polyseam/cndi/blob/main/CODE_OF_CONDUCT.md)

- [x ] I agree to follow this CNDI's Code of Conduct

# Notes (Optional)
## Node Pool Labeling:
Labels can now be specified for each node pool. This allows for more fine-grained control over the scheduling and placement of pods based on specific labels.
## Node Pool Tainting:
Taints are added to prevent certain pods from scheduling on nodes unless they tolerate the taints. This is crucial for nodes with specialized hardware or for separating certain workloads.
